go/common/crypto/signature/signers/ledger/ledger_signer.go

```diff
@@ -0,0 +1,126 @@
+package ledger
+
+import (
+	"errors"
+	"io"
+
+	"github.com/oasislabs/oasis-core/go/common/crypto/signature"
+
+	ledger "github.com/zondax/ledger-oasis-go"
+)
+
+const (
+	// SignerName is the name used to identify this package.
+	SignerName = "ledger"
+)
+
+var (
+	_ signature.SignerFactoryCtor = NewFactory
+	_ signature.SignerFactory     = (*Factory)(nil)
+	_ signature.Signer            = (*Signer)(nil)
+
+	// SignerDerivationPath is the derivation path used for generating
+	// the signature key on the Ledger device.
+	SignerDerivationPath = []uint32{44, 118, 0, 0, 0}
+
+	roleDerivationPaths = map[signature.SignerRole][]uint32{
+		signature.SignerEntity: SignerDerivationPath,
+	}
+)
+
+// Factory is a Ledger backed SignerFactory
+type Factory struct {
+	roles   []signature.SignerRole
+	address string
+}
+
+// NewFactory creates a new factory with the specified roles
+func NewFactory(address string, roles ...signature.SignerRole) signature.SignerFactory {
+	return &Factory{
+		roles:   append([]signature.SignerRole{}, roles...),
+		address: address,
+	}
+}
+
+// EnsureRole ensures that the SignatureFactory is configured for the given
+// role.
+func (fac *Factory) EnsureRole(role signature.SignerRole) error {
+	for _, v := range fac.roles {
+		if v == role {
+			return nil
+		}
+	}
+	return signature.ErrRoleMismatch
+}
+
+// Generate has the same functionality as Load, since all keys are generated
+// on the Ledger device.
+func (fac *Factory) Generate(role signature.SignerRole, _rng io.Reader) (signature.Signer, error) {
+	return fac.Load(role)
+}
+
+// Load will create a Signer backed by a Ledger device by searching for
+// a device with the expected address. If no address is provided, this will
+// created a Signer with the first Ledger device it finds.
+// The only role allowed is SignerEntity.
+func (fac *Factory) Load(role signature.SignerRole) (signature.Signer, error) {
+	path, ok := roleDerivationPaths[role]
+	if !ok {
+		return nil, errors.New("ledger signer is not allowed to have this role")
+	}
+
+	if fac.address == "" {
+		device, err := ledger.FindLedgerOasisApp()
+		if err != nil {
+			return nil, err
+		}
+
+		return &Signer{device, path}, nil
+	}
+
+	device, err := ledger.ConnectLedgerOasisApp(fac.address, path)
```

Comment:
Build will fail until https://github.com/ZondaX/ledger-oasis-go/pull/4, which enables ledger devices to be connected by address, is merged. We can't rely on a device index, since the index is not guaranteed to be the same between calls.

Reply:
What kind of address is this? This should be something that is independent of the derivation path and only dependent on the actual device (like a device ID).

Reply:
So from talking to Juan, there doesn't seem to be a unique ID to identify each device, so the solution is to use the address derived from each device's mnemonic.

Reply:
Should then this be based on a specific (static) derivation path, separate from the one used for the actual key? As otherwise if you have multiple roles (currently not, but we must think ahead), you will need to configure the address for each role which seems annoying?

Reply:
Right, so there are a few options that I see for resolving this issue, and would like to address it now. We could:
1. Leave it as is, the ListDevices command will always display the devices based on some static path, and the user will always used that path when connecting to the device, but will use a different path based on role.
2. We make the path configurable, adding a flag, and removing the role to path mapping. Or keep it as default paths. In which case, the path (or role) will be required for all calls involving Ledger. 

I'd prefer the first option since it would make the process for the users easier, although it does not allow for paths to be configurable.

Reply:
There is also option 3:
* Use a static path for listing and "pinning" devices.
* Use a (separate from the above) static path prefix based on role (e.g., specific role has specific prefix).
* Allow the user to configure the path suffix (and use a default if not specified).

Reply:
Ok that makes sense. I'll implement option 3 then.